### PR TITLE
JDBC driver returns 0 for precision and display size for projected character value expressions

### LINKED ISSUE #6975: JDBC driver returns 0 for precision and displaysize for projected character value expressions
Various vendors will return a precision and display size for the projected column and expression applied to the column. Presto is returning no precision or display size information for a column reference or character value expression. 

**Presto JDBC**
ColumnIndex	getColumnName	getColumnTypeName	getPrecision	getScale	isNullable	getTableName	getSchemaName	getCatalogName	getColumnClassName	getColumnDisplaySize	getColumnLabel	getColumnType	isAutoIncrement	isCaseSensitive	isCurrency	isDefinitelyWritable	isReadOnly	isSearchable	isSigned	isWritable
1	cchar	CHAR(32)	0	0	2				java.lang.String	0	cchar	1	false	false	false	false	true	true	false	false
2	_col1	VARCHAR(32)	0	0	2				java.lang.String	0	_col1	-16	false	false	false	false	true	true	false	false

**Teradata Presto JDBC**

ColumnIndex	getColumnName	getColumnTypeName	getPrecision	getScale	isNullable	getTableName	getSchemaName	getCatalogName	getColumnClassName	getColumnDisplaySize	getColumnLabel	getColumnType	isAutoIncrement	isCaseSensitive	isCurrency	isDefinitelyWritable	isReadOnly	isSearchable	isSigned	isWritable
1	cchar	CHAR(32)	32	0	1				java.lang.String	32	cchar	1	false	false	false	false	true	true	false	false
2	_col1	VARCHAR(32)	32	0	1				java.lang.String	32	_col1	12	false	false	false	false	true	true	false	false

Describe of referenced table

Column	Type	Comment
rnum	integer	
cchar	char(32)

### PR DESCRIPTION
**Pull request by delding**
FIX #6975 
Also reset "signed" meta info to be false for columns such as "timestamp", "date" and etc.

### CODE REVIEW

**Comment by electrum** [comment]
The commit message title is too long. Please see this guide: http://chris.beams.io/posts/git-commit/

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/ColumnInfo.java [hunk 1]
```diff
@@ -85,15 +85,16 @@ public ColumnInfo(
         this.catalogName = requireNonNull(catalogName, "catalogName is null");
     }
 
-    public static void setTypeInfo(Builder builder, TypeSignature type)
+    public static void setTypeInfo(Builder builder, TypeSignature typeSignature)
```
Comment (electrum) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/ColumnInfo.java]: Renaming this variable makes the diff harder to read. Such refactorings, if necessary, should be done as a separate commit to be easier to review and understand when looking at history.

Keep the original name and name the one below `typeName`

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/ColumnInfo.java [hunk 2]
```diff
@@ -182,10 +183,27 @@ public static void setTypeInfo(Builder builder, TypeSignature type)
...
+                builder.setScale(typeSignature.getParameters().get(1).getLongLiteral().intValue());
                 break;
+            default:
+                if (type.startsWith("char(") && type.endsWith(")")) {
+                    int size = Integer.parseInt(type.substring("char(".length(), type.length() - 1));
+                    builder.setSigned(false);
+                    builder.setPrecision(size);
+                    builder.setScale(0);
+                    builder.setColumnDisplaySize(size);
+                    break;
```
Comment (electrum) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/ColumnInfo.java]: Remove the breaks from default and make this `if ... else if ...`

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/ColumnInfo.java [hunk 3]
```diff
@@ -182,10 +183,27 @@ public static void setTypeInfo(Builder builder, TypeSignature type)
...
                 builder.setSigned(true);
-                builder.setColumnDisplaySize(type.getParameters().get(0).getLongLiteral().intValue() + 2); // dot and sign
-                builder.setPrecision(type.getParameters().get(0).getLongLiteral().intValue());
-                builder.setScale(type.getParameters().get(1).getLongLiteral().intValue());
+                builder.setColumnDisplaySize(typeSignature.getParameters().get(0).getLongLiteral().intValue() + 2); // dot and sign
+                builder.setPrecision(typeSignature.getParameters().get(0).getLongLiteral().intValue());
+                builder.setScale(typeSignature.getParameters().get(1).getLongLiteral().intValue());
                 break;
+            default:
+                if (type.startsWith("char(") && type.endsWith(")")) {
```
Comment (electrum) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/ColumnInfo.java]: This should be using `getBase()` on `TypeSignature`

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/ColumnInfo.java [hunk 4]
```diff
@@ -182,10 +183,27 @@ public static void setTypeInfo(Builder builder, TypeSignature type)
...
+                    int size = Integer.parseInt(type.substring("char(".length(), type.length() - 1));
```
Comment (electrum) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/ColumnInfo.java]: Use `getParameters()`

**Comment by delding** [comment]
@electrum , pr updated :-)

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/ColumnInfo.java [hunk 5]
```diff
@@ -186,6 +188,18 @@ public static void setTypeInfo(Builder builder, TypeSignature type)
                 builder.setPrecision(type.getParameters().get(0).getLongLiteral().intValue());
                 builder.setScale(type.getParameters().get(1).getLongLiteral().intValue());
                 break;
+            default:
```
Comment (raghavsethi) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/ColumnInfo.java]: You can use type.getBase instead of putting all these in the default branch.

Comment (delding) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/ColumnInfo.java]: addressed

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/ColumnInfo.java [hunk 6]
```diff
@@ -134,10 +137,28 @@ public static void setTypeInfo(Builder builder, TypeSignature type)
                 builder.setColumnDisplaySize(24);
                 break;
             case "varchar":
-                builder.setSigned(true);
-                builder.setPrecision(VARCHAR_MAX);
+                if ((parameters.size() == 1) && parameters.get(0).isLongLiteral()) {
```
Comment (electrum) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/ColumnInfo.java]: We don't need these checks -- the signature is guaranteed for the data type. See how decimal is done.

**Comment by raghavsethi** [comment]
Closing this, as it still needs some work.

**Comment by the6campbells** [comment]
will the fix be in 0.171?

**Comment by raghavsethi** [comment]
No, this needs to be re-done.

**Comment by the6campbells** [comment]
so which defect # is this tracked by ? Mine appears closed based on this which is closed.

**Comment by raghavsethi** [comment]
Sorry, I don't understand what you mean - #6975 is still open.
